contributors/design-proposals/api-machinery/client-go-context.md

```diff
@@ -0,0 +1,141 @@
+and also to allow exposing context based http tracing [go blog on context based httptrace](https://blog.golang.org/http-tracing) to consumers of the client.
+
+## Motivation and Goals
+
+When using client-go, external calls to the Kubernetes API are made in order to manage resources. The initiator of these calls may find due to an external event, 
+that they no longer need the resource they had requested, but they have no means by which to inform the client of this change. 
+
+When using client-go, having the ability to add context based http tracing would be valuable for issues around performance, observability and debugging. 
+
+- Allow consumers of the client to indicate, in an idiomatic way, that the action that caused them to invoke the client has been cancelled.
+- Allow consumers of the client to instrument and gain insight via context based http tracing.
+
+## Non Goals
+
+- Introduce a transactional system for write operations whereby upon cancellation the client would somehow rollback any current write operations.
+
+## API
+
+There are two approaches outlined. The first is a non breaking change that appends to the existing API for resources but perhaps puts more burden on the client consumer. 
+The second is a breaking change to the client-go api, that is potentially more idiomatic and clear.
+ 
+Common to both is the fact that the request type in client-go  already exposes a ```Context```  function: [request context set on client-go rest.Request](https://github.com/kubernetes/client-go/blob/master/rest/request.go#L393). 
+If set, this context is passed all the way through to the underlying ```http.Request```: [request context passed to http request code](https://github.com/kubernetes/client-go/blob/master/rest/request.go#L484). 
+So to achieve the goals, we need only modify the client API to allow passing in an external context that is then used to set the existing context on the request type.
+
+### Non breaking change
+
+The client resources (Pods, Secrets etc...) each expose a resource specific interface ```PodInterface``` for example. This is returned from a ```Getter``` interface:
+
+```go
+type PodsGetter interface {
+	Pods(namespace string) PodInterface
+}
+``` 
+
+If we were to modify these resource interfaces and their underlying concrete types ```pod``` in this case, to add a ```WithContext``` method, it would allow consumers of the client to set context while maintaining backwards compatibility for all other consumers of the API.
+
+```go
+// PodInterface has methods to work with Pod resources.
+type PodInterface interface {
+  	WithContext(ctx context.Context) PodInterface
+    ...
+}
+
+type pods struct {
+	client rest.Interface
+	ns     string
+	ctx    context.Context
+}
+
+// WithContext allows you to set a context that will be used by the underlying http request
+func (c *pods) WithContext(ctx context.Context) PodInterface {
+  	c.ctx = ctx
```

Comment:
`WithContext` should not mutate, but has to create a shallow copy first.

Reply:
Could you explain a little more, do you mean before the assignment on line 60 happens? My understanding is ctx is being passed by value into this method and so is a shallow copy or am I misunderstanding?

Reply:
I think I meant a shallow copy of `c *pods`.